docs/proposals/20220414-runtime-hooks.md

```diff
@@ -0,0 +1,426 @@
+
+The ability to orchestrate add-ons in line with events in the cluster lifecycle is becoming a requirement for many CAPI users, but in order to make this possible a mechanism to plug into the cluster lifecycle is required. This proposal introduces a set of Runtime Hooks designed to meet the need for add-on management orchestration including:
+
+* Operations for installing add-ons during the cluster provisioning workflow
+* Operations for upgrading add-ons during the cluster upgrade workflow
+* Operations for handling add-ons during the cluster deletion workflow
+
+Runtime Hooks enable the Cluster Lifecycle to trigger these processes based on the current state of the cluster, allowing them to start after some state has been reached or waiting for them to complete before moving on with Cluster-wide operations such as Cluster creation or deletion.
+
+Once these hooks are in place, it will be possible to build a comprehensive add-on orchestration solution on top of Cluster API that can leverage external tools such as helm, kapp, ArgoCD, flux or eventually CRS as an alternative.
+
+The hooks defined in this proposal are going to model common events in the Cluster lifecycle and we assume that in the future they may be used for other use cases beyond add-on management. 
+
+### Goals
+
+* Identify a set of Runtime Hooks that enable management of the entire add-on lifecycle 
+* Define the OpenAPI specification of these Runtime Hooks
+* Document when the corresponding Runtime Extensions are called
+* Provide guidelines for developers implementing a corresponding Runtime Extension
+
+### Non-Goals
+
+* Defining all possible Runtime Hooks in Cluster API; this proposal defines only a subset of hooks required for add-on orchestration.
+* Define a full add-on management solution or define detailed steps for solving add-on related problems like CPI migration from in-tree to out-of-tree; this proposal is focused only in providing the foundational capabilities to do so. 
+
+### Future work
+
+* Improve the hooks defined in this proposal to better support use cases beyond add-on management 
+
+
+## Proposal
+
+This proposal adds a set of Runtime Hooks specifically designed for Cluster created from a ClusterClass and thus leveraging the idea of managed topology. 
+
+The main reason for this choice is because a managed topology has a set of capabilities required for lifecycle hooks implementation:
+
+* A managed topology has the overarching view of the entire Cluster (vs other components in CAPI which are limited to controlling single resources or a subset of them).
+* A managed topology already has the capability to control all resources in a Cluster, thus making it possible to orchestrate lifecycle workflows like e.g. upgrades.
+
+Please note that this is especially relevant because it allows treating the cluster lifecycle as a state machine no matter the underlying controller being based on level-triggered reconcile loops.
```

Comment:
suggestion to address https://github.com/kubernetes-sigs/cluster-api/pull/6418/files#r869608110

```suggestion
In practice, if we look at the six lifecycle hooks introduced by this proposal, we should recongnize that four of them cannot be implemented outside of the topology controller because there is no viable point to be used in an alternative (BeforeClusterCreate, BeforeClusterUpgrade, AfterControlPlaneUpgrade, and AfterClusterUpgrade). 

Also, by working in the topology controller it is possible to implement hooks allowing to block the cluster from transitioning from one state to another, which is a capability required to properly orchestrate the addon lifecycle.
```